### DISCUSSION TITLE
Instrumentation

### DISCUSSION THREAD

**Discussion by skumaravelan** [graphql-java/graphql-java#2527] [Q&A]
Hi 
I am trying to log a request using Instrumentation API. I would like to know, is there any API that I can use to find given 
request is Query  or Mutation or Subscription and List of argument and requested elements in a query .

Thanks
Senthil

**Comment by dugenkui03** [comment]
`Instrumentation.instrumentExecutionInput` may satisfied your demand, but it can't acquire the `Document` object which is parsed from query-string.

If you want to get `Document`:
- if you have already implemented `PreparsedDocumentProvider`, you can get `Document` object and query variables in `PreparsedDocumentProvider.getDocument`;
- otherwise, you can find them in `Instrumentation.instrumentDocumentAndVariables`,  **but this way won't get invalid query info**.

**Comment by bbakerman** [comment]
This is not complilable code but just to roughly show you some ideas

```
        @Override
public InstrumentationContext<ExecutionResult> beginExecuteOperation(InstrumentationExecuteOperationParameters parameters)            
    AccessLoggerState accessLoggerState = parameters.getInstrumentationState();

            String operationName = parameters.getExecutionContext().getOperationDefinition().getName();
            String safeQuery = PrivacySafeQuery.privacySafe(parameters.getExecutionContext().getDocument(), operationName);

            return CompletableFuture.completedFuture(new InstrumentationContext<>() {
                @Override
                public void onDispatched(CompletableFuture<ExecutionResult> result) {
                }

                @Override
                public void onCompleted(ExecutionResult result, Throwable t) {
                    log.info("{}", safeQuery);
                }
            });
        }
